src/acvp_kdf135_ikev1.c

```diff
@@ -268,16 +290,87 @@ ACVP_RESULT acvp_kdf135_ikev1_kat_handler (ACVP_CTX *ctx, JSON_Object *obj) {
         groupval = json_array_get_value(groups, i);
         groupobj = json_value_get_object(groupval);
 
-        hash_alg = (unsigned char *) json_object_get_string(groupobj, "hashAlg");
-        auth_method = (unsigned char *) json_object_get_string(groupobj, "authenticationMethod");
+        hash_alg_str = json_object_get_string(groupobj, "hashAlg");
+        if (!hash_alg) {
+            ACVP_LOG_ERR("Failed to include hashAlg");
+            return ACVP_MISSING_ARG;
+        }
+
+        /*
+         * Determine the hash algorithm.
+         */
+        if (strncmp(hash_alg_str, ACVP_STR_SHA_1, strlen(ACVP_STR_SHA_1)) == 0) {
+            hash_alg = ACVP_KDF135_SHA1;
+        } else if (strncmp(hash_alg_str, ACVP_STR_SHA2_224, strlen(ACVP_STR_SHA2_224)) == 0) {
+            hash_alg = ACVP_KDF135_SHA224;
+        } else if (strncmp(hash_alg_str, ACVP_STR_SHA2_256, strlen(ACVP_STR_SHA2_256)) == 0) {
+            hash_alg = ACVP_KDF135_SHA256;
+        } else if (strncmp(hash_alg_str, ACVP_STR_SHA2_384, strlen(ACVP_STR_SHA2_384)) == 0) {
+            hash_alg = ACVP_KDF135_SHA384;
+        } else if (strncmp(hash_alg_str, ACVP_STR_SHA2_512, strlen(ACVP_STR_SHA2_512)) == 0) {
+            hash_alg = ACVP_KDF135_SHA512;
+        } else {
+            ACVP_LOG_ERR("ACVP server requesting invalid hashAlg");
+            return ACVP_INVALID_ARG;
+        }
+
+        auth_method_str = json_object_get_string(groupobj, "authenticationMethod");
+        if (!auth_method_str) {
+            ACVP_LOG_ERR("Failed to include authenticationMethod");
+            return ACVP_MISSING_ARG;
+        }
+
+        /*
+         * Determine the authentication method.
+         */
+        if (strncmp(auth_method_str, ACVP_AUTH_METHOD_DSA_STR,
+                    strlen(ACVP_AUTH_METHOD_DSA_STR)) == 0) {
```

Comment:
strneln ? there are a few of these in this file :)

Reply:
Otherwise this all looks good to me :)

Reply:
Thanks! I'll double check the others in this file. Let me explain my reasoning behind using strlen here and see if you still want me to change it :)

Since the function is being used on a non-changing, hard-coded string (#define) we already know the length of the string and don't need to use strnlen. This is in contrast to user input where we don't have total control over the object and in that case need to verify the length with an upper bound in place to be safe. It's also good to also use strnlen for use with variables because the string value can also be expanded. However, since this string value is resolved by the pre-processor during compilation, it won't change during run-time.

@bfussell This is my current understanding. Am I making any incorrect assumptions here?

Thanks!
Philip

Reply:
I understand, and I thought about that too. I decided to comment because I think it would be good to just set precedence, you know? If we decide it isn't necessary then all good :) But I think it would be good to use those safe string functions everywhere if that is something we will care about in PRs in the future :)

Reply:
Apparently the other CSMs use strnlen even on hard coded strings, actually they use strnlen_s :)
In this case merge it as is for now. We'll need to go thru all the code to convert to the safeC APIs at some point, correct ?